File: frontend/src/types/interview.ts
export type InterviewStage =
  | "INTRO"
  | "CLARIFICATION"
  | "APPROACH_DISCUSSION"
  | "PSEUDOCODE"
  | "CODING"
  | "COMPLEXITY_DISCUSSION"
  | "FOLLOW_UP"
  | "FEEDBACK"
  | "COMPLETE";

export type InterviewRole = "system" | "assistant" | "user";
export type InterviewSessionStatus = "ACTIVE" | "COMPLETED" | "ABANDONED";

export interface InterviewSessionResponse {
  id: string;
  user_id: string;
  problem_id: string;
  stage: InterviewStage;
  status: InterviewSessionStatus;
  final_score?: number | null;
  stuck_signal_count: number;
  nudges_used_in_stage: number;
  started_at: string;
  completed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface InterviewMessageResponse {
  id: string;
  session_id: string;
  user_id?: string | null;
  role: InterviewRole;
  content: string;
  stage_at_message: InterviewStage;
  created_at: string;
}

export interface InterviewEvaluationResponse {
  id: string;
  session_id: string;
  stage: InterviewStage;
  problem_understanding_score: number;
  approach_quality_score: number;
  code_correctness_reasoning_score: number;
  complexity_analysis_score: number;
  communication_clarity_score: number;
  total_score: number;
  passed: boolean;
  summary?: string | null;
  rubric_json?: Record<string, unknown>;
  created_at: string;
}

export interface InterviewSessionDetailResponse extends InterviewSessionResponse {
  messages: InterviewMessageResponse[];
  evaluations: InterviewEvaluationResponse[];
  can_code?: boolean;
}

export interface InterviewCompletionResponse extends InterviewSessionResponse {
  strengths: string[];
  gaps: string[];
  next_steps: string[];
}

export interface StartInterviewSessionPayload {
  problem_id: string;
}

export interface InterviewMessagePayload {
  content: string;
  role?: "user";
  has_submission?: boolean;
  current_code?: string;
  chat_history?: Array<{ role: "user" | "assistant"; content: string }>;
}

File: backend/app/schemas/interview.py
from datetime import datetime
from typing import Any, Literal
from pydantic import BaseModel, Field


InterviewStage = Literal[
    "INTRO",
    "CLARIFICATION",
    "APPROACH_DISCUSSION",
    "PSEUDOCODE",
    "CODING",
    "COMPLEXITY_DISCUSSION",
    "FOLLOW_UP",
    "FEEDBACK",
    "COMPLETE",
]

InterviewSessionStatus = Literal["ACTIVE", "COMPLETED", "ABANDONED"]
InterviewMessageRole = Literal["system", "assistant", "user"]


class InterviewSessionCreate(BaseModel):
    problem_id: str


class InterviewSessionUpdateStage(BaseModel):
    stage: InterviewStage


class InterviewSessionComplete(BaseModel):
    final_score: float | None = Field(default=None, ge=0, le=50)


class InterviewMessageCreate(BaseModel):
    content: str
    role: InterviewMessageRole = "user"
    has_submission: bool = False
    current_code: str | None = None
    chat_history: list[dict[str, str]] = Field(default_factory=list)


class InterviewEvaluationCreate(BaseModel):
    stage: InterviewStage
    problem_understanding_score: int = Field(ge=0, le=10)
    approach_quality_score: int = Field(ge=0, le=10)
    code_correctness_reasoning_score: int = Field(ge=0, le=10)
    complexity_analysis_score: int = Field(ge=0, le=10)
    communication_clarity_score: int = Field(ge=0, le=10)
    total_score: float = Field(ge=0, le=50)
    passed: bool = False
    summary: str | None = None
    rubric_json: dict[str, Any] = Field(default_factory=dict)


class InterviewMessageResponse(BaseModel):
    id: str
    session_id: str
    user_id: str | None = None
    role: InterviewMessageRole
    content: str
    stage_at_message: InterviewStage
    created_at: datetime

    class Config:
        orm_mode = True


class InterviewEvaluationResponse(BaseModel):
    id: str
    session_id: str
    stage: InterviewStage
    problem_understanding_score: int
    approach_quality_score: int
    code_correctness_reasoning_score: int
    complexity_analysis_score: int
    communication_clarity_score: int
    total_score: float
    passed: bool
    summary: str | None = None
    rubric_json: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime

    class Config:
        orm_mode = True


class InterviewSessionResponse(BaseModel):
    id: str
    user_id: str
    problem_id: str
    stage: InterviewStage
    status: InterviewSessionStatus
    final_score: float | None = None
    stuck_signal_count: int = 0
    nudges_used_in_stage: int = 0
    started_at: datetime
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True


class InterviewSessionDetailResponse(InterviewSessionResponse):
    messages: list[InterviewMessageResponse] = Field(default_factory=list)
    evaluations: list[InterviewEvaluationResponse] = Field(default_factory=list)
    can_code: bool = False


class InterviewCompletionResponse(InterviewSessionResponse):
    strengths: list[str] = Field(default_factory=list)
    gaps: list[str] = Field(default_factory=list)
    next_steps: list[str] = Field(default_factory=list)
